### ISSUE TITLE
[ConnectionManager] Application stopped but CM still alive

### ISSUE THREAD

**Issue by Aprogiena** [stratisproject/StratisBitcoinFullNode#679]
This shouldn't happen:

```
info: Stratis.Bitcoin.Base.BaseFeature[0]
      Flushing headers chain
info: Stratis.Bitcoin.FullNode[0]
      Waiting for FlushChain to finish.

Application stopped.

info: Stratis.Bitcoin.Connection.ConnectionManager[0]
      Node [::ffff:51.141.28.47]:26178 offline
info: Stratis.Bitcoin.Connection.ConnectionManager[0]
      Reason: Unexpected exception while connecting to socket
```

**Comment by Aprogiena** [comment]
I think this issue is related to the crash that I've just experienced on shutdown:
 
```
[2017-11-03 08:05:30.5915 7] INFO: Stratis.Bitcoin.Utilities.AsyncLoop+<>c__DisplayClass12_0+<<StartAsync>b__0>d.MoveNext PeriodicLog stopping
[2017-11-03 08:05:30.5915 13] INFO: Stratis.Bitcoin.FullNode.Stop Closing node pending...
[2017-11-03 08:05:30.5915 7] INFO: Stratis.Bitcoin.Utilities.AsyncLoop+<>c__DisplayClass12_0+<<StartAsync>b__0>d.MoveNext BlockStore.DownloadAndStoreBlocks stopping
[2017-11-03 08:05:30.5915 7] INFO: Stratis.Bitcoin.Utilities.AsyncLoop+<>c__DisplayClass12_0+<<StartAsync>b__0>d.MoveNext PosMining.Mine stopping
[2017-11-03 08:05:30.5915 7] INFO: Stratis.Bitcoin.Utilities.AsyncLoop+<>c__DisplayClass12_0+<<StartAsync>b__0>d.MoveNext FlushAddressManager stopping
[2017-11-03 08:05:30.5915 7] INFO: Stratis.Bitcoin.Utilities.AsyncLoop+<>c__DisplayClass12_0+<<StartAsync>b__0>d.MoveNext MemoryPool.RelayWorker stopping
[2017-11-03 08:05:30.5915 7] INFO: Stratis.Bitcoin.Utilities.AsyncLoop+<>c__DisplayClass12_0+<<StartAsync>b__0>d.MoveNext FlushChain stopping
[2017-11-03 08:05:30.5915 7] INFO: Stratis.Bitcoin.Utilities.AsyncLoop+<>c__DisplayClass12_0+<<StartAsync>b__0>d.MoveNext wallet persist job stopping
[2017-11-03 08:05:30.5915 25] INFO: Stratis.Bitcoin.Connection.ConnectionManagerBehavior.AttachedNode_StateChanged Node [::ffff:104.219.251.201]:26178 offline
[2017-11-03 08:05:30.5915 27] INFO: Stratis.Bitcoin.Connection.ConnectionManagerBehavior.AttachedNode_StateChanged Node [::ffff:40.78.65.98]:26178 offline
[2017-11-03 08:05:30.5915 31] INFO: Stratis.Bitcoin.Connection.ConnectionManagerBehavior.AttachedNode_StateChanged Node [::ffff:84.192.155.226]:26178 offline
[2017-11-03 08:05:30.5915 50] INFO: Stratis.Bitcoin.Connection.ConnectionManagerBehavior.AttachedNode_StateChanged Node [::ffff:51.140.27.200]:26178 offline
[2017-11-03 08:05:30.5915 29] INFO: Stratis.Bitcoin.Connection.ConnectionManagerBehavior.AttachedNode_StateChanged Node [::ffff:86.22.243.246]:26178 offline
[2017-11-03 08:05:30.5915 33] INFO: Stratis.Bitcoin.Connection.ConnectionManagerBehavior.AttachedNode_StateChanged Node [::ffff:51.140.47.104]:26178 offline
[2017-11-03 08:05:30.5915 38] INFO: Stratis.Bitcoin.Connection.ConnectionManagerBehavior.AttachedNode_StateChanged Node [::ffff:13.64.76.48]:26178 offline
[2017-11-03 08:05:30.6071 13] INFO: Stratis.Bitcoin.Utilities.AsyncLoop.Dispose Waiting for PeriodicLog to finish.
[2017-11-03 08:05:30.6223 13] INFO: Stratis.Bitcoin.Utilities.AsyncLoop.Dispose Waiting for PosMining.Mine to finish.
[2017-11-03 08:05:30.6223 13] INFO: Stratis.Bitcoin.Utilities.AsyncLoop.Dispose Waiting for wallet persist job to finish.
[2017-11-03 08:05:33.8412 13] INFO: Stratis.Bitcoin.Features.MemoryPool.MempoolFeature.Stop Saving Memory Pool...
[2017-11-03 08:05:33.8412 13] INFO: Stratis.Bitcoin.Features.MemoryPool.MempoolFeature.Stop ...Memory Pool Saved 0 transactions
[2017-11-03 08:05:33.8412 13] INFO: Stratis.Bitcoin.Utilities.AsyncLoop.Dispose Waiting for MemoryPool.RelayWorker to finish.
[2017-11-03 08:05:33.8412 13] INFO: Stratis.Bitcoin.Features.BlockStore.BlockStoreFeature.Stop Stopping BlockStore...
[2017-11-03 08:05:33.8412 13] INFO: Stratis.Bitcoin.Utilities.AsyncLoop.Dispose Waiting for BlockStore.RelayWorker to finish.
[2017-11-03 08:05:33.8412 13] INFO: Stratis.Bitcoin.Utilities.AsyncLoop.Dispose Waiting for BlockStore.DownloadAndStoreBlocks to finish.
[2017-11-03 08:05:33.8773 13] INFO: Stratis.Bitcoin.Utilities.AsyncLoop.Dispose Waiting for Consensus Loop to finish.
[2017-11-03 08:05:33.8773 13] INFO: Stratis.Bitcoin.Features.Consensus.ConsensusFeature.Stop Flushing Cache CoinView...
[2017-11-03 08:05:33.9852 13] INFO: Stratis.Bitcoin.Base.BaseFeature.Stop Flushing address manager
[2017-11-03 08:05:33.9852 13] INFO: Stratis.Bitcoin.Utilities.AsyncLoop.Dispose Waiting for FlushAddressManager to finish.
[2017-11-03 08:05:33.9852 13] INFO: Stratis.Bitcoin.Base.BaseFeature.Stop Flushing headers chain
[2017-11-03 08:05:33.9852 13] INFO: Stratis.Bitcoin.Utilities.AsyncLoop.Dispose Waiting for FlushChain to finish.
[2017-11-03 08:06:57.1213 35] ERROR: Stratis.Bitcoin.Features.MemoryPool.MempoolBehavior+<AttachedNode_MessageReceivedAsync>d__26.MoveNext Exception occurred: System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'LoggerFactory'.
   at Microsoft.Extensions.Logging.LoggerFactory.CreateLogger(String categoryName)
   at Stratis.Bitcoin.Utilities.AsyncLoopFactory..ctor(ILoggerFactory loggerFactory) in C:\StratisBitcoinFullNode\Stratis.Bitcoin\Utilities\AsyncLoopFactory.cs:line 76
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(IServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitSingleton(SingletonCallSite singletonCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(IServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(IServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitSingleton(SingletonCallSite singletonCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(IServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(IServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitSingleton(SingletonCallSite singletonCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(IServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteExpressionBuilder.<>c__DisplayClass14_0.<Build>b__0(ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.GetService(Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetService[T](IServiceProvider provider)
   at Stratis.Bitcoin.FullNode.NodeService[T](Boolean failWithDefault) in C:\StratisBitcoinFullNode\Stratis.Bitcoin\FullNode.cs:line 91
   at Stratis.Bitcoin.Base.ChainState.get_IsInitialBlockDownload() in C:\StratisBitcoinFullNode\Stratis.Bitcoin\Base\ChainState.cs:line 72
   at Stratis.Bitcoin.Features.MemoryPool.MempoolBehavior.<ProcessInvAsync>d__29.MoveNext() in C:\StratisBitcoinFullNode\Stratis.Bitcoin.Features.MemoryPool\MempoolBehavior.cs:line 346
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Stratis.Bitcoin.Features.MemoryPool.MempoolBehavior.<AttachedNode_MessageReceivedAsync>d__26.MoveNext() in C:\StratisBitcoinFullNode\Stratis.Bitcoin.Features.MemoryPool\MempoolBehavior.cs:line 193
```

**Comment by Aprogiena** [comment]
I haven't gone too deeply, but for the first look it seems like a race condition.

Consider Node.cs and its BeginListen there is 
```
_Cancel = new CancellationTokenSource();
```

so the cancellation token source is only initialized after BeginListen is started. But there is plenty of space before that call when someone might want to invoke shutdown.

For example in `internal Node(NetworkAddress peer, Network network, NodeConnectionParameters parameters)` we have `State = NodeState.Connected;`. So once this is executed then `Node.IsConnect` returns true. And there is a big gap before BeginListen is called. So if DisconnectAll is called which leads to DisconnectAsync then this can lead to crash or whatever behavior including this.

Solution? Bring Node to FN and rewrite it. Get rid of its threading design, make it fully async, get rid of SocketAsyncEventArgs.

**Comment by dangershony** [comment]
Yes looks like we will take the approach of moving Protocol over and refactor once its ported lets dicuss more in tomorrows call.

**Comment by fassadlr** [comment]
@Aprogiena @noescape00 @noescape00 guys Im pretty sure that #825 fix this bug?

**Comment by Aprogiena** [comment]
will also keep an eye on it

**Comment by fassadlr** [comment]
This issue was fixed with #825. If something like this  occurs again, please create a new issue.
